{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import shutil\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output='/node05/program/jiachenghao/Mycoplasmoides_pneumoniae/Qulity/mNGS/'\n",
    "# input_path='/node05/program/jiachenghao/Mycoplasmoides_pneumoniae/肺炎支原体测序数据/Transfer/'\n",
    "# input_path_list=os.listdir(input_path)\n",
    "# file=pd.read_excel('QC.xlsx')['File'].to_list()\n",
    "\n",
    "# for i in file:\n",
    "#     copy_file=[x for x in input_path_list if x.startswith(i)]\n",
    "#     for ii in copy_file:\n",
    "#          shutil.copy(input_path+ii,output)\n",
    "    # if len(copy_file) == 1:\n",
    "    #     copy_file=copy_file[0]\n",
    "    # shutil.copy(input_path+copy_file,output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Cal_overage(genome_coverage_txt):\n",
    "    global coverage_percentage\n",
    "    coverage_data = []\n",
    "    with open('genome_coverage.txt', 'r') as file:\n",
    "        for line in file:\n",
    "            parts = line.strip().split()\n",
    "            reference, depth, num_bases, total_bases, fraction = parts\n",
    "            if reference == 'genome':\n",
    "                depth = int(depth)\n",
    "                num_bases = int(num_bases)\n",
    "                total_bases = int(total_bases)\n",
    "                fraction = float(fraction)\n",
    "                coverage_data.append((depth, num_bases, total_bases, fraction))\n",
    "    total_bases = coverage_data[0][2]\n",
    "    total_coverage = sum(depth * num_bases for depth, num_bases, _, _ in coverage_data)\n",
    "    average_coverage = total_coverage / total_bases\n",
    "    # print(f\"Average Coverage: {average_coverage:.2f}\")\n",
    "    covered_bases = total_bases - next(num_bases for depth, num_bases, _, _ in coverage_data if depth == 0)\n",
    "    coverage_percentage = (covered_bases / total_bases) * 100\n",
    "    # print(f\"Coverage Percentage for : {coverage_percentage:.2f}%\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "output='/node05/program/jiachenghao/Mycoplasmoides_pneumoniae/Qulity/test/'   ######\n",
    "Script_path='/node05/program/jiachenghao/Mycoplasmoides_pneumoniae/Qulity/'\n",
    "workpath='/node05/program/jiachenghao/Mycoplasmoides_pneumoniae/Qulity/coverage/'\n",
    "output_list=os.listdir(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['UDB-143']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Process_file=os.listdir(output)\n",
    "Single_list=[]\n",
    "for i in Process_file:\n",
    "    i=i.split('_')[2]\n",
    "    Single_list.append(i)\n",
    "Single_list=list(set(Single_list))\n",
    "Single_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Process  UDB-143\n",
      "Bowtie2 DB Run Success\n",
      "Settings:\n",
      "  Output files: \"refgenome.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  NC_000912.1.fasta\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 204098\n",
      "Using parameters --bmax 153074 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 153074 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 816394 (target: 153073)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 816394 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 816395 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 249211\n",
      "fchr[G]: 412131\n",
      "fchr[T]: 575834\n",
      "fchr[$]: 816394\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4466719 bytes to primary EBWT file: refgenome.1.bt2\n",
      "Wrote 204104 bytes to secondary EBWT file: refgenome.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 816394\n",
      "    bwtLen: 816395\n",
      "    sz: 204099\n",
      "    bwtSz: 204099\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 51025\n",
      "    offsSz: 204100\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 4253\n",
      "    numLines: 4253\n",
      "    ebwtTotLen: 272192\n",
      "    ebwtTotSz: 272192\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 204098\n",
      "Using parameters --bmax 153074 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 153074 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 816394 (target: 153073)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 816394 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 816395 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 249211\n",
      "fchr[G]: 412131\n",
      "fchr[T]: 575834\n",
      "fchr[$]: 816394\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4466719 bytes to primary EBWT file: refgenome.rev.1.bt2\n",
      "Wrote 204104 bytes to secondary EBWT file: refgenome.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 816394\n",
      "    bwtLen: 816395\n",
      "    sz: 204099\n",
      "    bwtSz: 204099\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 51025\n",
      "    offsSz: 204100\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 4253\n",
      "    numLines: 4253\n",
      "    ebwtTotLen: 272192\n",
      "    ebwtTotSz: 272192\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:01\n",
      "\n",
      "Step2 Run Success\n",
      "\n",
      "bam file is ready\n",
      "Step3 Run Success\n",
      "\n",
      "Genome_coverage file is ready\n"
     ]
    }
   ],
   "source": [
    "File=[]\n",
    "Coverage=[]\n",
    "for i in Single_list:\n",
    "    File.append(i)\n",
    "    Check=os.listdir(workpath)\n",
    "    if len(Check) != 0:\n",
    "        print('Folder not Empty')\n",
    "    print('Start Process ', i)\n",
    "    \n",
    "    index1=i+'_1'\n",
    "    index2=i+'_2'\n",
    "    file1=[x for x in output_list if str(index1) in x][0]\n",
    "    file2=[x for x in output_list if str(index2) in x][0]\n",
    "\n",
    "    shutil.copy(output+file1,workpath)\n",
    "    shutil.copy(output+file2,workpath)\n",
    "    shutil.copy(Script_path+'NC_000912.1.fasta',workpath)\n",
    "    shutil.copy(Script_path+'step2.sh',workpath)\n",
    "    shutil.copy(Script_path+'step3.sh',workpath)\n",
    "##################################################################\n",
    "    command = [\"bowtie2-build\", \"NC_000912.1.fasta\", \"refgenome\"]\n",
    "    result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=workpath)\n",
    "    if result.returncode == 0:\n",
    "        print(\"Bowtie2 DB Run Success\")\n",
    "        print(result.stdout.decode())\n",
    "    else:\n",
    "        print(\"Bowtie2 DB Run Failed\")\n",
    "        print(result.stderr.decode())\n",
    "################################################################\n",
    "    command = [\"sh\", \"step2.sh\"]\n",
    "    result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=workpath)\n",
    "    if result.returncode == 0:\n",
    "        print(\"Step2 Run Success\")\n",
    "        print(result.stdout.decode())\n",
    "    else:\n",
    "        print(\"Step2 Run Failed\")\n",
    "        print(result.stderr.decode())\n",
    "###############################################################\n",
    "    if os.path.isfile(workpath+'mapping_result_sorted.bam'):\n",
    "        print('bam file is ready')\n",
    "    else:\n",
    "        print('Error')\n",
    "###############################################################\n",
    "    command = [\"sh\", \"step3.sh\"]\n",
    "    result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=workpath)\n",
    "    if result.returncode == 0:\n",
    "        print(\"Step3 Run Success\")\n",
    "        print(result.stdout.decode())\n",
    "    else:\n",
    "        print(\"Step3 Run Failed\")\n",
    "        print(result.stderr.decode())\n",
    "###############################################################\n",
    "    if os.path.isfile(workpath+'genome_coverage.txt'):\n",
    "        print('Genome_coverage file is ready')\n",
    "    else:\n",
    "        print('Error')\n",
    "\n",
    "\n",
    "    shutil.copy(workpath+'genome_coverage.txt',Script_path)\n",
    "    Cal_overage('genome_coverage.txt')\n",
    "    Coverage.append(coverage_percentage)\n",
    "\n",
    "    os.remove(Script_path+'genome_coverage.txt')\n",
    "    delet=os.listdir(workpath)\n",
    "    for i in delet:\n",
    "        os.remove(workpath+i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['UDB-143']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[94.14265660943123]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "out={'File':File,'Coverage':Coverage}\n",
    "out=pd.DataFrame(out)\n",
    "out.to_csv('mNGS_Coverage.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "shenzhen",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
